src/Pos/Security/Workers.hs

```diff
@@ -0,0 +1,186 @@
+        throwM $ DBMalformed $
+            sformat ("Eclipse check: didn't manage to find parent of "%build%
+                     " with hash "%build%", which is not genesis")
+                    (headerHash header)
+                    (header ^. prevBlockL)
+    -- We stop looking for blocks when we've gone earlier than
+    -- 'mdNoBlocksSlotThreshold':
+    pastThreshold header =
+        (flattenSlotId slotId - flattenEpochOrSlot header) >
+        mdNoBlocksSlotThreshold
+    -- Run the iteration starting from tip block; if we have found
+    -- that we're eclipsed, we report it and ask neighbors for
+    -- headers. If there are no main blocks generated by someone else
+    -- in the past 'mdNoBlocksSlotThreshold' slots, it's bad and we've
+    -- been eclipsed.  Here's how we determine that a block is good
+    -- (i.e. main block generated not by us):
+    isGoodBlock (Left _)   = False
+    isGoodBlock (Right mb) = mb ^. headerLeaderKey /= ourPk
+    -- Okay, now let's iterate until we see a good blocks or until we
+    -- go past the threshold and there's no point in looking anymore:
+    notEclipsed header = do
+        let prevBlock = header ^. prevBlockL
+        if | pastThreshold header     -> pure False
+           | prevBlock == genesisHash -> pure True
+           | isGoodBlock header       -> pure True
+           | otherwise                ->
+                 getBlockHeader prevBlock >>= \case
+                     Just h  -> notEclipsed h
+                     Nothing -> onBlockLoadFailure header $> True
+
+checkForReceivedBlocksWorkerImpl
+    :: forall ssc m.
+       (SscWorkersClass ssc, WorkMode ssc m)
+    => SendActions m -> m ()
+checkForReceivedBlocksWorkerImpl sendActions = afterDelay $ do
+    repeatOnInterval (const (sec' 4)) . reportingFatal version $
+        whenM (needRecovery $ Proxy @ssc) $ do
+            triggerRecovery sendActions
+    repeatOnInterval (min (sec' 20)) . reportingFatal version $ do
+        ourPk <- ncPublicKey <$> getNodeContext
+        let onSlotDefault slotId = do
+                header <- getTipBlockHeader @ssc
+                unlessM (checkEclipsed ourPk slotId header) onEclipsed
+        maybe (pure ()) onSlotDefault =<< getCurrentSlot
+  where
+    sec' :: Int -> Millisecond
+    sec' = convertUnit . sec
+    afterDelay action = delay (sec 3) >> action
+    onEclipsed = do
+        logWarning $
+            "Our neighbors are likely trying to carry out an eclipse attack! " <>
+            "There are no blocks younger " <>
+            "than 'mdNoBlocksSlotThreshold' that we didn't generate " <>
+            "by ourselves"
+        reportEclipse
+    repeatOnInterval delF action = runIfNotShutdown $ do
+        () <- action
+        getLastKnownSlotDuration >>= delay . delF
+        repeatOnInterval delF action
+    reportEclipse = do
+        bootstrapMin <- (+ sec 10) . convertUnit <$> getLastKnownSlotDuration
+        nonTrivialUptime <- (> bootstrapMin) <$> getUptime
```

Comment:
I believe that the inaccuracies of using `UTCTime` instead of a monotonic time may introduce e.g. 500ms differences in time. `getUptime` is calculated based on `now` and `start`: if there was a clock change inbetween now and star, then that could yield a `500ms` difference.

It might not be a big deal as we're comparing a 10s range, on the other hand: why not be sure.